File: c/2015/day03a/day03a.c
#include <stdio.h>
#include <stdlib.h>

#include "string_set.h"

typedef struct position {
    int x;
    int y;
} position_t;

void usage(char *progname) {
    fprintf(stderr, "usage: %s <input file>\n", progname);
    exit(1);
}

int process(char *filename) {
    FILE *infile;
    char ch;
    string_set_t *positions = string_set_new();
    position_t santa = { 0, 0 };
    char buf[16];
    int size;
  
    if ( (infile = fopen(filename, "r")) == NULL ) {
        perror("failed to open input file");
        exit(1);
    }
    while ( !feof(infile) ) {
        ch = (char)fgetc(infile);
        switch ( ch ) {
            case '^':
                santa.y += 1;
                break;
            case 'v':
                santa.y -= 1;
                break;
            case '<':
                santa.x -= 1;
                break;
            case '>':
                santa.x += 1;
                break;
        }
        sprintf(buf, "%d,%d", santa.x, santa.y);
        string_set_insert(positions, buf);
    }
    fclose(infile);
    size = string_set_size(positions);
    string_set_free(positions);
    return size;
}

int main(int argc, char *argv[]) {
    char *filename;
    int result;

    if ( argc < 2 ) {
        usage(argv[0]);
    }

    filename = argv[1];
    result = process(filename);
    printf("result = %d\n", result);
    return 0;
}

File: pascal/2015/day03a/day03a.pas
PROGRAM Day03a(Input, Output);

USES
    Generics.Collections;

TYPE
    Position2D = RECORD
        X, Y : 	Integer
    END;

    PositionSet = SPECIALIZE THashSet<Position2D>;

VAR
    FileName : PChar;
    Res      : UInt32;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Process(FileName : PChar) : UInt32;
VAR
    InFile    : Text;
    Ch        : Char;
    Santa     : Position2D;
    Positions : PositionSet;
    Count     : UInt32;

BEGIN
    Santa.X := 0;
    Santa.Y := 0;
    Positions := PositionSet.Create;
    Positions.Add(Santa);
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        Read(InFile, Ch);
        CASE Ch OF
            '^' : Santa.Y := Santa.Y + 1;
            'v' : Santa.Y := Santa.Y - 1;
            '<' : Santa.X := Santa.X - 1;
            '>' : Santa.X := Santa.X + 1;
        END;
        Positions.Add(Santa);
    END;
    Close(InFile);
    Count := Positions.GetCount;
    Positions.Free;
    Process := Count
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.
